TST: Enable coveralls/codecov code coverage

jenshnielsen opened the pull request:
This is a rebase of #4367 on top of #4668 / #4188 

#4668 should be merges before this one. 

Changes are: 

* Use pip install -e . to install a develop build. This should fix the issue with paths
* Use setup.py test from #4188 to run tests and generate coverage

jenshnielsen commented:
Forget it. The relative paths only work with the pep8 tests that we don't want

jenshnielsen commented:
Looks like codecov.io works much better. For those interested it is running on my master branch here https://codecov.io/github/jenshnielsen/matplotlib 

I still have to do some fine tuning before this is ready to merge

mdboom commented:
Just curious -- I've only ever used coveralls, but have found it to be unresponsive at times.  Any other advantages to codecov.io?

jenshnielsen commented:
I spent a significant amount of time at the Scipy sprint trying to get the relative/absolute links to the source working correctly with coveralls. The answer is probably burried somewhere in our setup/installation but with codecov it just seems to work. 

There are other issues at the moment. I had to manually exclude the source code of numpy etc. which should already be excluded.

mdboom commented:
@jenshnielsen: This is cool -- I'd like to start improving coverage, and this is a logical first step.  What do you think remains to be done here?  I have no problem going with codecov over coveralls if that's working better.

jenshnielsen commented:
AFAIK it's basically working. I decided to postpone it till after 1.5 but I should be ready to get back to it. Let me see if I can rebase it over the weekend.

mdboom commented on tests.py:
```diff
@@ -23,6 +23,10 @@
 
+    env = {"NOSE_WITH_COVERAGE": 1,
+           'NOSE_COVER_PACKAGE': 'matplotlib',
+           'NOSE_COVER_HTML': 1}
```
This should add to, not replace, `os.environ`.

mdboom replied:
Since code coverage can slow down runtimes, and it's not compatible with multiprocessing tests, we should probably have this be an option controlled by a commandline flag.  Then on Travis-CI, we can do this just on one configuration (probably Python 3.5), and keep all the others running with multiple processes.

mdboom commented on tests.py:
```diff
@@ -33,8 +37,8 @@ def run(extra_args):
-              argv=sys.argv + extra_args)
-
+              argv=sys.argv + extra_args,
+              env=env)
```
This should also be done in `lib/matplotlib/__init__.py:test`.  Probably move the definition of the environment variables to a private function in `__init__.py` and then get them from here and the `test` function.

mdboom commented on .travis.yml:
```diff
@@ -100,7 +103,7 @@ script:
     if [[ $BUILD_DOCS == false ]]; then
       export MPL_REPO_DIR=$PWD  # needed for pep8-conformance test of the examples
-      gdb -return-child-result -batch -ex r -ex bt --args python tests.py -s --processes=$NPROC --process-timeout=300 $TEST_ARGS
+      gdb -return-child-result -batch -ex r -ex bt --args python tests.py $TEST_ARGS
```
I think we should do one config with coverage, the rest with multiprocess.  multiprocess catches a certain class of bugs that I'd hate to stop catching.

jenshnielsen replied:
Yes. At this stage I was just experimenting to see it that is the issue. The code coverage is build against my branch at https://codecov.io/github/jenshnielsen/matplotlib/commits unfortunately I still have issues getting it to work correctly. Either the links to files don't work or they are not there at all.

mdboom commented on .travis.yml:
```diff
@@ -89,8 +89,11 @@ install:
       cp tmp/usr/share/fonts/truetype/humor-sans/Humor-Sans.ttf ~/.fonts
       cp Felipa-Regular.ttf ~/.fonts
       fc-cache -f -v
+      python setup.py install
```
Could the doc build do `pip install .` too?  Or does that not work?

jenshnielsen replied:
It doesn't work at the moment because the docs needs the rc templates file copied by the setup.py build step. We can fix that but at this stage it's not even clear the the `pip install .` is the right solution. The idea is that it will install into a predictable dir (because it does a flat install) that does not contain the hash of the release and can be configured in codecov or coveralls

mdboom replied:
Thanks.

mdboom commented on .travis.yml:
```diff
@@ -161,3 +164,6 @@ after_success:
       fi
   fi
+  if [[ $BUILD_DOCS == false && $TEST_ARGS != "--pep8-only" ]]; then
+     codecov
```
Mostly just curious about this: How does this uploading work without credentials?  Is anyone able to just upload matplotlib coverage results?

mdboom commented:
This is looking quite cool.  I assume when all is said and done, there is a way to not print the coverage report to the console on Travis?  It makes it hard to scroll up to find any actual failures or errors.

jenshnielsen commented:
I have made a new personal fork of nose with @tacaswell's changes as well as merging the branch that adds support for not displaying the output to stdout

jenshnielsen commented:
I think this should finally be ready to go. I guess the main question is if we want to do this or wait for the py.test work which will change this and hopefully make it a lot simpler.

I have moved the environment to __init__.py and made it possible to run coverage from matplotlib.test() via a function argument. I have reverted to coveralls which seems more stabile than codecov. 

The only way have been able to make either of codecov and coveralls able to figure out the files is to do a develop install. This also means that the log is a lot shorted because al the files are not copied. 
I have done that via `pip install -e .` which means that no log at all is shown on a successful build. 

Furthermore this is running my fork of nose to be able to suppress the output to std out. 

According to the nose docs and various issues you only get reliable coverage output when you disable multiprocessing so I have done that on one of the jobs and only run coverage on that.

mdboom commented:
With respect for waiting for py.test -- I don't see a strong reason to wait given that you've already done the work here.  But I'll leave this up for a while in case there are differing opinions.  I agree it should hopefully be easier under py.test in the long run.

tacaswell commented:
I am :+1: on merging this now.

WeatherGod commented:
Just noticed that the mpl_toolkits aren't "covered".

WeatherGod commented:
I should clarify that what I mean is that the mpl_toolkits, while tested (somewhat), aren't included in any metrics or reporting.

tacaswell commented:
`cbook.py` also has some really funny looking coverage results.

jenshnielsen commented:
@WeatherGod I don't think mpl_toolkits are included in @tacaswell original work. I think we just need to add it to the list of modules covered.

jenshnielsen commented:
Yes there is something odd in cbook and other places. Functions that are not recorded as called but have lines within them registered as called. I will try to reproduce locally.

tacaswell commented:
I am going to merge this now and iterate on it as we go.
